prql/src/semantic/materializer.rs

```diff
@@ -0,0 +1,586 @@
+//! Transform the parsed AST into a "materialized" AST, by executing functions and
+//! replacing variables. The materialized AST is "flat", in the sense that it
+//! contains no query-specific logic.
+use std::collections::HashMap;
+
+use crate::ast::*;
+use crate::ast_fold::*;
+use crate::error::{Error, Reason};
+
+use anyhow::{bail, Result};
+use itertools::zip;
+use itertools::Itertools;
+
+use super::scope::Context;
+use super::scope::TableColumn;
+
+pub struct SelectedColumns(pub Vec<Node>);
+
+/// Replaces all resolved functions and variables with their declarations.
+pub fn materialize(
+    nodes: Vec<Node>,
+    context: Context,
+) -> Result<(Vec<Node>, Context, SelectedColumns)> {
```

Comment:
Something I'm thinking about, which doesn't need immediate action, is "how easy it is for a new contributor to understand the code and try out simple examples at a high level"

If we can keep an "easy case" so that someone can take a PRQL string, push it through some functions, and get the result, that would be great.

This code does a lot, which is awesome. One change this makes is to add a couple of steps to the pipeline; e.g. `materialize(ast)` is now:

```
        let (res, context) = resolve(ast.nodes, None)?;
        let (mat, _, _) = materialize(res.clone(), context)?;
```

Possibly there's some work we can do with Default structs, or accepting `None` and making a default in those cases.

Again — no need to change things now, let's keep it in our mind though. There's a natural progression towards making things more complicated as we add features and people become more accustomed to the code, so it's often an area we need to push ourselves to think about.

Reply:
Awesome, thank you!

I definitely agree it's a hard problem — the existing ones were easy, and so it's easy to find good abstractions. But now we're getting to the parts where the abstractions break down a bit.

One tiny point, if that's OK: function names like `process` aren't that descriptive without context; so if we can name them something a bit more distinct, that makes it easier for to discuss the code and for new contributors to find their way around. The `semantic` module is a good example of this done well. Sometimes that leads us to ugly names (`materialize` 😬)!

Reply:
True that. Do you have a suggestion for the name?